File: 78.subsets.java
/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ret = new ArrayList<>();
        for (int ix = 0; ix != (int) Math.pow(2, nums.length); ++ix) {
            List<Integer> lst = new ArrayList<>();
            for (int i = 0; i != nums.length; ++i) {
                if ((ix & (0x01 << i)) != 0x00) {
                    lst.add(nums[i]);
                }
            }
            ret.add(lst);
        }
        return ret;
    }
}
// @lc code=end

File: 78.subsets.cpp
/*
 * @lc app=leetcode id=78 lang=cpp
 *
 * [78] Subsets
 */

// @lc code=start

#include <vector>

using namespace std;

class Solution {
private:
    vector<vector<int>> results_;
    vector<int> bag_;
    vector<int> candidates_;

    void backtracking(const vector<int>::iterator &pos) {
        results_.push_back(bag_);
        if (pos == candidates_.end()) {
            return;
        }
        for (auto iter = pos; iter != candidates_.end(); ++iter) {
            bag_.push_back(*iter);
            backtracking(iter + 1);
            bag_.pop_back();
        }
    }

public:
    vector<vector<int>> subsets(vector<int>& nums) {
        candidates_ = nums;
        backtracking(candidates_.begin());
        return results_;
    }
};
// @lc code=end
